presto-native-execution/presto_cpp/main/connectors/arrow_flight/ArrowFlightConnector.cpp

```diff
@@ -0,0 +1,184 @@
+class CallOptionsAddHeaders : public FlightCallOptions, public AddCallHeaders {
+ public:
+  void AddHeader(const std::string& key, const std::string& value) override {
+    headers.emplace_back(key, value);
+  }
+};
+
+std::optional<Location> ArrowFlightConnector::getDefaultLocation(
+    const std::shared_ptr<FlightConfig>& config) {
+  auto defaultHost = config->defaultServerHostname();
+  auto defaultPort = config->defaultServerPort();
+  if (!defaultHost.has_value() || !defaultPort.has_value()) {
+    return std::nullopt;
+  }
+
+  bool defaultSslEnabled = config->defaultServerSslEnabled();
+  AFC_RETURN_OR_RAISE(
+      defaultSslEnabled
+          ? Location::ForGrpcTls(defaultHost.value(), defaultPort.value())
+          : Location::ForGrpcTcp(defaultHost.value(), defaultPort.value()));
+}
+
+std::shared_ptr<arrow::flight::FlightClientOptions>
+ArrowFlightConnector::initClientOpts(
+    const std::shared_ptr<FlightConfig>& config) {
+  auto clientOpts = std::make_shared<FlightClientOptions>();
+  clientOpts->disable_server_verification = !config->serverVerify();
+
+  auto certPath = config->serverSslCertificate();
+  if (certPath.has_value()) {
+    std::ifstream file(certPath.value());
+    VELOX_CHECK(file.is_open(), "Could not open TLS certificate");
+    std::string cert(
+        (std::istreambuf_iterator<char>(file)),
+        (std::istreambuf_iterator<char>()));
+    clientOpts->tls_root_certs = cert;
+  }
+
+  return clientOpts;
+}
+
+FlightDataSource::FlightDataSource(
+    const RowTypePtr& outputType,
+    const std::unordered_map<std::string, std::shared_ptr<ColumnHandle>>&
+        columnHandles,
+    std::shared_ptr<auth::Authenticator> authenticator,
+    memory::MemoryPool* pool,
+    const std::shared_ptr<FlightConfig>& flightConfig,
+    const std::shared_ptr<arrow::flight::FlightClientOptions>& clientOpts,
+    const std::optional<arrow::flight::Location>& defaultLocation)
+    : outputType_{outputType},
+      authenticator_{std::move(authenticator)},
+      pool_{pool},
+      flightConfig_{flightConfig},
+      clientOpts_{clientOpts},
+      defaultLocation_{defaultLocation} {
+  // columnMapping_ contains the real column names in the expected order.
+  // This is later used by projectOutputColumns to filter out unnecessary
+  // columns from the fetched chunk.
+  columnMapping_.reserve(outputType_->size());
+
+  for (const auto& columnName : outputType_->names()) {
+    auto it = columnHandles.find(columnName);
+    VELOX_CHECK(
+        it != columnHandles.end(),
+        "missing columnHandle for column '{}'",
+        columnName);
+
+    auto handle = std::dynamic_pointer_cast<FlightColumnHandle>(it->second);
+    VELOX_CHECK_NOT_NULL(
+        handle,
+        "handle for column '{}' is not an FlightColumnHandle",
+        columnName);
+
+    columnMapping_.push_back(handle->name());
+  }
+}
+
+void FlightDataSource::addSplit(std::shared_ptr<ConnectorSplit> split) {
```

Comment:
Did some minor cleanup in this method to simply use of the serialized FlightEndpoint